![title](wcar.png)
# Polysomal mRNA Association and Gene Expression in Trypanosoma brucei
## This repository contains the scripts used for the data analysis
    Michele Tinti Anna Kernel and Michael A.J. Ferguson

    Wellcome Centre for Anti-Infectives Research
    School of Life Sciences, University of Dundee

The Jupyter notebook
- Figures_Paper_def.ipynb
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/mtinti/polysome/master?filepath=Figures_Paper_def.ipynb)

produces the figures and tables submitted for the paper

![Alt text](Figures/Figure_1.jpeg?raw=true "Title")

The contrasting physiological environments of Trypanosoma brucei procyclic (insect vector) and bloodstream (mammalian host) forms necessitates deployment of different  molecular processes and, therefore, changes in protein expression. Transcriptional regulation is unusual in T. brucei because the arrangement of genes is polycistronic; however, genes which are transcribed together are subsequently cleaved into separate mRNAs by trans-splicing. Following pre-mRNA processing, the regulation of mature mRNA stability is a tightly controlled cellular process. While many stage-specific transcripts have been identified, previous studies using RNA-seq suggest that changes in overall transcript level do not necessarily reflect the abundance of the corresponding protein. To better understand the regulation of gene expression in T. brucei, we performed RNA-seq on total, sub-polysomal, and polysomal mRNA. Our analyses also showed that several long non-coding RNAs are more abundant in the sub-polysome samples, suggesting they might pay a role in the regulation of gene expression. 
